improve short-circuiting on (any|drop)null(!)

cjprybol opened the pull request:
motivation
This PR implements logic tests for `anynull` to avoid checking arrays that cannot contain `Nullable` elements. This PR supersedes https://github.com/JuliaStats/NullableArrays.jl/pull/188 by also fixing `dropnull` and `dropnull!` by extending the same logic tests to account for `Vector{Union}`s, which are handled incorrectly by the current `dropnull` and `dropnull!`

benchmarks and tests
```julia
using NullableArrays, BenchmarkTools, Base.Test
srand(1); x = rand(1_000_000); @benchmark anynull(x)
srand(1); x = rand(1000, 1000); @benchmark anynull(x)
srand(1); x = convert(Array{Union{Int, Float64}}, rand(1_000_000)); @benchmark anynull(x)
srand(1); x = convert(Array{Union{Int, Float64}}, rand(1000, 1000)); @benchmark anynull(x)
x = convert(Array{Union{Int, Nullable{Int}}}, repmat([1, Nullable()], 5)); dropnull!(x)
x = convert(Array{Union{Int, Nullable{Int}}}, repmat([1, Nullable()], 5)); dropnull(x)
x = convert(Array{Union{Int, Nullable{Int}, Float64, Nullable{Float64}}}, repmat([1, Nullable()], 5)); dropnull!(x)
x = convert(Array{Union{Int, Nullable{Int}, Float64, Nullable{Float64}}}, repmat([1, Nullable()], 5)); dropnull(x)
```

results
````julia
# benchmark
srand(1); x = rand(1_000_000); @benchmark anynull(x)
# PR
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     17.377 ns (0.00% GC)
  median time:      17.669 ns (0.00% GC)
  mean time:        19.026 ns (0.00% GC)
  maximum time:     81.627 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     998
  time tolerance:   5.00%
  memory tolerance: 1.00%
# master
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     270.970 μs (0.00% GC)
  median time:      283.878 μs (0.00% GC)
  mean time:        318.548 μs (0.00% GC)
  maximum time:     1.278 ms (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%


############################################################################################
# benchmark
srand(1); x = rand(1000, 1000); @benchmark anynull(x)
# PR
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     16.856 ns (0.00% GC)
  median time:      17.410 ns (0.00% GC)
  mean time:        18.608 ns (0.00% GC)
  maximum time:     94.667 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     998
  time tolerance:   5.00%
  memory tolerance: 1.00%
# master
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     270.965 μs (0.00% GC)
  median time:      286.472 μs (0.00% GC)
  mean time:        311.942 μs (0.00% GC)
  maximum time:     1.024 ms (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%


############################################################################################
# benchmark
srand(1); x = convert(Array{Union{Int, Float64}}, rand(1_000_000)); @benchmark anynull(x)
# PR
BenchmarkTools.Trial:
  memory estimate:  128 bytes
  allocs estimate:  2
  --------------
  minimum time:     443.242 ns (0.00% GC)
  median time:      488.109 ns (0.00% GC)
  mean time:        513.604 ns (1.91% GC)
  maximum time:     10.368 μs (92.00% GC)
  --------------
  samples:          10000
  evals/sample:     198
  time tolerance:   5.00%
  memory tolerance: 1.00%
# master
BenchmarkTools.Trial:
  memory estimate:  15.26 MiB
  allocs estimate:  1000000
  --------------
  minimum time:     31.832 ms (0.00% GC)
  median time:      34.032 ms (1.36% GC)
  mean time:        34.662 ms (1.02% GC)
  maximum time:     42.589 ms (0.00% GC)
  --------------
  samples:          144
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%

############################################################################################
# benchmark
srand(1); x = convert(Array{Union{Int, Float64}}, rand(1000, 1000)); @benchmark anynull(x)
# PR
BenchmarkTools.Trial:
  memory estimate:  128 bytes
  allocs estimate:  2
  --------------
  minimum time:     445.313 ns (0.00% GC)
  median time:      481.639 ns (0.00% GC)
  mean time:        514.056 ns (1.98% GC)
  maximum time:     11.591 μs (90.53% GC)
  --------------
  samples:          10000
  evals/sample:     198
  time tolerance:   5.00%
  memory tolerance: 1.00%
# master
BenchmarkTools.Trial:
  memory estimate:  15.26 MiB
  allocs estimate:  1000000
  --------------
  minimum time:     12.168 ms (0.00% GC)
  median time:      14.276 ms (4.17% GC)
  mean time:        14.866 ms (3.17% GC)
  maximum time:     28.467 ms (3.35% GC)
  --------------
  samples:          335
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%


############################################################################################
# test
x = convert(Array{Union{Int, Nullable{Int}}}, repmat([1, Nullable()], 5)); dropnull!(x)
# PR
5-element Array{Int64,1}:
 1
 1
 1
 1
 1
# master
10-element Array{Union{Int64, Nullable{Int64}},1}:
 1
 #NULL
 1
 #NULL
 1
 #NULL
 1
 #NULL
 1
 #NULL

############################################################################################
# test
x = convert(Array{Union{Int, Nullable{Int}}}, repmat([1, Nullable()], 5)); dropnull(x)
# PR
5-element Array{Int64,1}:
 1
 1
 1
 1
 1
# master
10-element Array{Union{Int64, Nullable{Int64}},1}:
 1
 #NULL
 1
 #NULL
 1
 #NULL
 1
 #NULL
 1
 #NULL

############################################################################################
# test
x = convert(Array{Union{Int, Nullable{Int}, Float64, Nullable{Float64}}}, repmat([1, Nullable()], 5)); dropnull!(x)
# PR
5-element Array{Union{Float64, Int64},1}:
 1
 1
 1
 1
 1
# master
10-element Array{Union{Float64, Int64, Nullable{Float64}, Nullable{Int64}},1}:
 1
 #NULL
 1
 #NULL
 1
 #NULL
 1
 #NULL
 1
 #NULL


############################################################################################
# test
x = convert(Array{Union{Int, Nullable{Int}, Float64, Nullable{Float64}}}, repmat([1, Nullable()], 5)); dropnull(x)
# PR
5-element Array{Union{Float64, Int64},1}:
 1
 1
 1
 1
 1
# master
10-element Array{Union{Float64, Int64, Nullable{Float64}, Nullable{Int64}},1}:
 1
 #NULL
 1
 #NULL
 1
 #NULL
 1
 #NULL
 1
 #NULL
```

cjprybol committed:
make allnull and anynull no-ops if eltypes are not Nullable

cjprybol committed:
revert allnull changes as that short-circuits quickly enough as is

cjprybol committed:
use multiple dispatch rather than ternary

cjprybol committed:
update anynull to handle more cases and add tests

cjprybol committed:
add version-specific checks for union types

cjprybol committed:
needed to hit command-Z one more time...

cjprybol committed:
add static to version check

cjprybol committed:
move version check/union function to top level and cleanup tests

cjprybol committed:
update anynull and dropnull to both short circuit when the array type
can inform us that no nulls are present

cjprybol committed:
add ternary to handle case of Union eltypes when creating similar array

cjprybol committed:
remove Nullable types from the Union return array

nalimilan commented on src/primitives.jl:
```diff
@@ -161,11 +167,18 @@ unwrapping `Nullable` entries. A copy is always returned, even when
 """
 function dropnull{T}(X::AbstractVector{T})                  # -> AbstractVector
-    if !(Nullable <: T) && !(T <: Nullable)
+    u = isa(T, Union)
```
Can you move this logic to a `isnullable` function? That will avoid code duplication and will also allow testing directly for corner cases like unions and `Any`.

nalimilan commented on src/primitives.jl:
```diff
@@ -153,6 +153,12 @@ end
 _isnull(x::Any) = false
 _isnull(x::Nullable) = isnull(x)
 
+if VERSION < v"0.6.0-dev.2107"
```
I know @ararslan took the time to check the commit version, but in general it's simpler and better to just do `isdefined(Base, :uniontypes)`.

nalimilan commented on test/primitives.jl:
```diff
@@ -358,5 +358,48 @@ module TestPrimitives
+        @test !anynull(Nullable.(rand(10, 10)))
+        @test !anynull(Array{Union{Nullable{Int}, Int}}(collect(1:10)))
+        @test anynull(Array{Union{Nullable{Int}, Int}}(rand([1, Nullable()], 100)))
+        @test !anynull(NullableArray{Union{Nullable{Int}, Int}}(collect(1:10)))
```
`NullableArray` with a union of nullable and non-nullable entries doesn't sound like a common (or even meaningful) use case. Rather test with a `Union{Int, Float64}` eltype.

nalimilan commented on test/primitives.jl:
```diff
@@ -358,5 +358,48 @@ module TestPrimitives
 
+    @testset "dropnull" begin
+        x = fill(1, 10)
+        @test dropnull!(x) == x && dropnull!(x) === x
```
`===` implies `==`.

nalimilan commented on test/primitives.jl:
```diff
@@ -358,5 +358,48 @@ module TestPrimitives
+        @test dropnull(Nullable.(x)) == x
+        @test dropnull!(NullableArray(x)) == x
+        @test dropnull(NullableArray(x)) == x
+        @test dropnull!(convert(Array{Union{Int, Nullable{Int}}}, x)) == x
```
Test the return type?

cjprybol replied:
added another test set to check return types

nalimilan commented on test/primitives.jl:
```diff
@@ -358,5 +358,48 @@ module TestPrimitives
+        @test dropnull(x) == x && !(dropnull(x) === x)
+        x = repmat([1, Nullable()], 5)
+        @test isequal(dropnull!(copy(x)), fill(1, 5))
+        @test isequal(dropnull(copy(x)), fill(1, 5))
```
No need for `copy!`.

nalimilan commented on test/primitives.jl:
```diff
@@ -358,5 +358,48 @@ module TestPrimitives
+        @test isequal(dropnull(copy(x)), fill(1, 5))
+        @test isequal(dropnull!(NullableArray(x)), fill(1, 5))
+        @test isequal(dropnull(NullableArray(x)), fill(1, 5))
+        @test isequal(dropnull!(convert(Array{Union{Int, Nullable{Int}}}, x)), fill(1, 5))
```
Also test the return type here.

nalimilan commented on src/primitives.jl:
```diff
@@ -182,11 +195,18 @@ unwrapped `Nullable` entries. If no nulls are present, this is a no-op
         deleteat!(X, find(isnull, X))
-        res = similar(X, eltype(T))
+        if u
+            res = similar(X, Union{filter(t -> !(t <: Nullable), _uniontypes(T))...})
```
You shouldn't filter out nullable types, since that means you won't be able to store some values in the result, e.g. if the input was `Union{Int, Nullable{String}}`. Need to call `eltype` instead. This needs to be tested.

Also use upper case letters like `U` for types.

nalimilan replied:
You shouldn't filter out nullable types, since that means you won't be able to store some values in the result, e.g. if the input was `Union{Int, Nullable{String}}`. Need to call `eltype` instead. This needs to be tested. Also use upper case letters like `U` for types.

This should also go to a separate function, you really want to avoid any code duplication, especially for this kind of tricky checks.

cjprybol committed:
fix Union type handling, extend testing, move duplicated checks to functions

Was filtering out Nullable{T} elements from Unions. That was incorrect,
and now Nullable{T}s are replaced by their unwrapped T type in the
return Union type. A few edits for existing tests, and extended tests to
cover new functions and the return type for dropnull and dropnull!.

nalimilan commented:
Actually, we may not even need the `anynull` changes: see https://github.com/JuliaLang/julia/issues/21256.

cjprybol commented:
That'd be a great improvement! Would you like to hold off on this PR and see how things play out with that issue? I could remove the anynull changes and leave the dropnull fixes.

nalimilan commented:
Actually, @mbauman just gave me a much simpler way to check whether a type can hold a `Nullable`: `typeintersect(T, Nullable) !== Union{}`. So we can simplify `_isnullable` a lot.

nalimilan commented on src/primitives.jl:
```diff
@@ -210,8 +238,15 @@ dropnull!(X::NullableVector) = deleteat!(X, find(X.isnull)).values # -> Vector
-anynull(X::NullableArray) = any(X.isnull)   # -> Bool
+anynull(X::Any) = any(_isnull, X)                              # -> Bool
+anynull(X::NullableArray) = any(X.isnull)                      # -> Bool
+function anynull{T}(X::AbstractArray{T})                       # -> Bool
```
We won't even need this if we fix https://github.com/JuliaLang/julia/issues/21256. Doesn't sound too hard, and would be a better way of spending our efforts.

ararslan commented:
Something we did in DataArrays that you might want to consider here is deprecating `allnull(x)` and `anynull(x)` in favor of `all(isnull, x)` and `any(isnull, x)`, respectively. The same level of specialization can still occur; you just overload e.g. `all(::typeof(isnull), x::NullableArray)`.

cjprybol commented:
I like both of those suggestions, closing in favor of https://github.com/JuliaStats/NullableArrays.jl/pull/192. I'll use mbauman's much better `_isnullable` in the linked DataTables PR. Thanks all!
